cognite/client/_api_client.py

```diff
@@ -263,57 +263,115 @@ def _list_generator(
+                    if sort is not None:
+                        body["sort"] = sort
+                    res = self._post(url_path=resource_path + "/list", json=body, headers=headers)
+                else:
+                    raise ValueError("_list_generator parameter `method` must be GET or POST, not {}".format(method))
+                last_received_items = res.json()["items"]
+                total_items_retrieved += len(last_received_items)
+
+                if not chunk_size:
+                    for item in last_received_items:
+                        yield cls._RESOURCE._load(item, cognite_client=self._cognite_client)
+                else:
+                    current_items.extend(last_received_items)
+                    if len(current_items) >= chunk_size:
+                        items_to_yield = current_items[:chunk_size]
+                        current_items = current_items[chunk_size:]
+                        yield cls._load(items_to_yield, cognite_client=self._cognite_client)
+
+                next_cursor = res.json().get("nextCursor")
+                if total_items_retrieved == limit or next_cursor is None:
+                    if chunk_size and current_items:
+                        yield cls._load(current_items, cognite_client=self._cognite_client)
+                    break
+
+    def _list_generator_partitioned(
+        self,
+        partitions: int,
+        cls,
+        resource_path: str,
+        filter: Dict = None,
+        other_params: Dict = None,
+        headers: Dict = None,
+    ):
+        next_cursors = {i + 1: None for i in range(partitions)}
+
+        def get_partition(partition_num):
+            next_cursor = next_cursors[partition_num]
+
+            body = {
+                "filter": filter or {},
+                "limit": self._LIST_LIMIT,
+                "cursor": next_cursor,
+                "partition": "{}/{}".format(partition_num, partitions),
+                **(other_params or {}),
+            }
+            res = self._post(url_path=resource_path + "/list", json=body, headers=headers)
             next_cursor = res.json().get("nextCursor")
-            if total_items_retrieved == limit or next_cursor is None:
-                if chunk_size and current_items:
-                    yield cls._load(current_items, cognite_client=self._cognite_client)
-                break
+            next_cursors[partition_num] = next_cursor
+
+            return res.json()["items"]
+
+        while len(next_cursors) > 0:
+            tasks_summary = utils._concurrency.execute_tasks_concurrently(
+                get_partition, [(partition,) for partition in next_cursors], max_workers=partitions
+            )
+            if tasks_summary.exceptions:
+                raise tasks_summary.exceptions[0]
+
+            for item in tasks_summary.joined_results():
+                yield cls._RESOURCE._load(item, cognite_client=self._cognite_client)
```

Comment:
I don't see the point of this, it's not returning as it's retrieving, so has no benefit over for a in c.bla.list()

Reply:
It yields in-between the partitions so only a single page pr partition is read before starting to return, reducing memory usage quite substantially for very large sets of data

Reply:
aha I get it now, that's a nice solution!